// ra-ui-materialui/dist/esm/field/ReferenceOneField.js
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
import React, { useMemo } from 'react';
import { Typography } from '@mui/material';
import { useReferenceOneFieldController, useRecordContext, ResourceContextProvider, useCreatePath, useTranslate, ReferenceFieldContextProvider, } from 'ra-core';
import { ReferenceFieldView } from './ReferenceField';
/**
 * Render the related record in a one-to-one relationship
 *
 * Expects a single field as child
 *
 * @example // display the bio of the current author
 * <ReferenceOneField reference="bios" target="author_id">
 *     <TextField source="body" />
 * </ReferenceOneField>
 */
export var ReferenceOneField = function (props) {
    var _a;
    var children = props.children, reference = props.reference, _b = props.source, source = _b === void 0 ? 'id' : _b, target = props.target, emptyText = props.emptyText, sort = props.sort, filter = props.filter, _c = props.link, link = _c === void 0 ? false : _c, queryOptions = props.queryOptions;
    var record = useRecordContext(props);
    var createPath = useCreatePath();
    var translate = useTranslate();
    var controllerProps = useReferenceOneFieldController({
        record: record,
        reference: reference,
        source: source,
        target: target,
        sort: sort,
        filter: filter,
        queryOptions: queryOptions,
    });
    var resourceLinkPath = !record || link === false
        ? false
        : createPath({
            resource: reference,
            id: (_a = controllerProps.referenceRecord) === null || _a === void 0 ? void 0 : _a.id,
            type: typeof link === 'function'
                ? link(record, reference)
                : link,
        });
    var context = useMemo(function () { return (__assign(__assign({}, controllerProps), { link: resourceLinkPath })); }, [controllerProps, resourceLinkPath]);
    return !record ||
        (!controllerProps.isPending &&
            controllerProps.referenceRecord == null) ? (emptyText ? (React.createElement(Typography, { component: "span", variant: "body2" }, emptyText && translate(emptyText, { _: emptyText }))) : null) : (React.createElement(ResourceContextProvider, { value: reference },
        React.createElement(ReferenceFieldContextProvider, { value: context },
            React.createElement(ReferenceFieldView, { reference: reference, source: source }, children))));
};
// disable sorting on this field by default as its default source prop ('id')
// will match the default sort ({ field: 'id', order: 'DESC'})
// leading to an incorrect sort indicator in a datagrid header
ReferenceOneField.sortable = false;
