// ra-core/dist/cjs/controller/field/useReferenceFieldController.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.useReferenceFieldController = void 0;
var react_1 = require("react");
var routing_1 = require("../../routing");
var useReference_1 = require("../useReference");
var core_1 = require("../../core");
var util_1 = require("../../util");
var useReferenceFieldController = function (options) {
    var _a = options.link, link = _a === void 0 ? 'edit' : _a, reference = options.reference, queryOptions = options.queryOptions;
    if (!reference) {
        throw new Error('useReferenceFieldController: missing reference prop. You must provide a reference, e.g. reference="posts".');
    }
    var id = (0, util_1.useFieldValue)(options);
    var referenceRecordQuery = (0, useReference_1.useReference)({
        reference: reference,
        id: id,
        options: __assign(__assign({}, queryOptions), { enabled: ((queryOptions === null || queryOptions === void 0 ? void 0 : queryOptions.enabled) == null ||
                (queryOptions === null || queryOptions === void 0 ? void 0 : queryOptions.enabled) === true) &&
                id != null }),
    });
    var createPath = (0, routing_1.useCreatePath)();
    var resourceDefinition = (0, core_1.useResourceDefinition)({ resource: reference });
    var result = (0, react_1.useMemo)(function () {
        return (__assign(__assign({}, referenceRecordQuery), { link: referenceRecordQuery.referenceRecord != null
                ? link === false ||
                    (link === 'edit' && !resourceDefinition.hasEdit) ||
                    (link === 'show' && !resourceDefinition.hasShow)
                    ? false
                    : createPath({
                        resource: reference,
                        id: referenceRecordQuery.referenceRecord.id,
                        type: typeof link === 'function'
                            ? link(referenceRecordQuery.referenceRecord, reference)
                            : link,
                    })
                : undefined }));
    }, [createPath, link, reference, referenceRecordQuery, resourceDefinition]);
    return result;
};
exports.useReferenceFieldController = useReferenceFieldController;
